### ISSUE TITLE
Fixture definition rework

### ISSUE THREAD

**Issue by mcallegari** [mcallegari/qlcplus#1026]
The plan is [documented here](https://docs.google.com/document/d/1cnHt-ITpa1Uk-cJD6HgGHqy0Dl1bobEHYOGgf8zaJkc/edit?usp=sharing)

Items on the list:

- [ ] Channel presets
- [ ] Capability presets
- [ ] Global/per-mode physical information
- [ ] Channel alias

**Comment by mcallegari** [comment]
@janosvitok @siegmund42 @fxedel for reference

**Comment by fxedel** [comment]
Will you change the channel alias syntax as well to support multiple modes?

**Comment by mcallegari** [comment]
> Will you change the channel alias syntax as well to support multiple modes?

I don't think so. I believe you're referring to the example Lukas gave about the Eurolite bar.
When assigning DMX addresses to a Fixture, you need to avoid conflicts with other addresses.
So, the Eurolite bar will have always 12 channels reserved, where by default when CH1 = 0, the other 11 channels will be set to "Nothing". 
When changing CH1 value, aliases will kick in, and CH2 to CH12 will appear/disappear on the fly.

In general, guys, please let's not complicate things too much. Everything could be done, indeed, but ask yourself: is it worth it ?
Since this rework is affecting the engine, I don't want to end up handling super complicated cases used maybe by 2-3 fixtures over 1000.
Please let's be practical.

**Comment by mcallegari** [comment]
Feature 4 addressed by a18c1242be9bcc648c6b600a0bfcd17da537ad2b

**Comment by fxedel** [comment]
I don't refer to Lukas' example with different modes depending on the first channel. I refer to the reuse of such channels in several (normal, traditional) modes.

My problem is that in your example, the channel number `7` is hardcoded:
````xml
<Capability Min="200" Max="255" Preset=”Alias”>Sound active
  <Alias name=”Sound sensitivity” number=”7” />
</Capability>
````

This may work in mode 1, where Sound Sensitivity is at position 7, but not in mode 2, where Sound Sensitivity is at position 9 (e.g. because Pan fine and Tilt fine are added in this mode). This syntax simply doesn't allow the reuse in different modes – and I thought you were aiming to reduce redundancy?

My proposal was to not reference a channel number (`7`), but a channel placeholder string, like `Program Speed / Sound Sensitivity`: The capability sets the value of this placeholder to an existent channel (like `Sound Sensitivity`) and the mode only references the placeholder.

[That's the way how we handle it at OFL](https://github.com/FloEdelmann/open-fixture-library/blob/master/docs/fixture-format.md#switching-channels), and it allows us to reuse such switching channels amongst different modes.

Here's a list of example fixtures – they all use the same switching channel in multiple modes at different positions, which means that a hardcoded channel number wouldn't work:
- [American DJ Revo Burst](https://open-fixture-library.herokuapp.com/american-dj/revo-burst)
- [cameo Hydrabeam 100](https://open-fixture-library.herokuapp.com/cameo/hydrabeam-100)
- [cameo Hydrabeam 300 RGBW](https://open-fixture-library.herokuapp.com/cameo/hydrabeam-300-rgbw)
- [Futurelight DMH-75.i LED](https://open-fixture-library.herokuapp.com/futurelight/dmh-75-i-led-moving-head)
- [Martin MAC Viper AirFX](https://open-fixture-library.herokuapp.com/martin/mac-viper-airfx)
- [Martin MAC Viper Performance](https://open-fixture-library.herokuapp.com/martin/mac-viper-performance)
- [Robe ColorSpot 2500E AT](https://open-fixture-library.herokuapp.com/robe/colorspot-2500e-at)
- [Showtec LED Light Bar RGB V3](https://open-fixture-library.herokuapp.com/showtec/led-light-bar-rgb-v3)
- [Stairville Stage TRI LED](https://open-fixture-library.herokuapp.com/stairville/stage-tri-led)
- [Venue TriStrip3Z](https://open-fixture-library.herokuapp.com/venue/tristrip3z)

As of practicalness: Of 109 OFL fixtures, 28 (25.6%) use switching channels and 10 (9.2%) of them use them in different modes in different positions – Lukas' fixture is an edge case, but this isn't.

**Comment by mcallegari** [comment]
@fxedel I see your point, but I'm not convinced about placeholders. It's too difficult for users.
I need to think some more about it.
Maybe my idea just lacks of a "mode" attribute, like this:
```
<Capability Min="200" Max="255" Preset=”Alias”>Sound active
  <Alias name=”Sound sensitivity” mode="10 Channel" number=”7” />
</Capability>
```
So basically the workflow is:
- define all the possible channels supported by a fixture
- add modes
- define aliases

**Comment by mcallegari** [comment]
Feature 2 addressed by 72aa60259f95aa6cdf840d05a0f2394377d11bcd

**Comment by fxedel** [comment]
@mcallegari I haven't thought about this solution, but this fixes the mode-issue as well. Nice! Looking forward to exporting our switching channels to QLC+ aliases without data loss 😃 

Could you give an example how defaultValue would be implemented in XML? We already store it in OFL, so it might be pretty easy to export this on our side.

**Comment by mcallegari** [comment]
> Could you give an example how defaultValue would be implemented in XML? We already store it in OFL, so it might be pretty easy to export this on our side.

Pretty straightforward actually:
```xml
<Channel Name="Tilt" Default="128" Preset="PositionTilt"/>
```

**Comment by siegmund42** [comment]
> Lukas' fixture is an edge case

At least it is part of a group of "edge cases". I did some research and it turns out at least the following fixtures already included in QLC+ will most probably have the same problem:

- Stairville LED BAR 252 RGB (no. 234564
- Chauvet 4Bar and 4Bar Tri
- Eurolite LED Bar RGB 252/10 40° Indoor (as already mentioned)

Some fixtures not yet included in QLC+ with possible same problem:

- Chauvet ColorBar
- Eurolite LED Bar-18 TCL

I don't exactly know the implications of this, but I guess the number of fixtures is quite low and nobody has ever complained about a problem with it so it should be fine (let's hope no manufacturer will built such fixtures with a silly DMX chart). Just was curious how much of the fixtures will be affected.

Anyway, keep up the good work - looks good to me so far!

**Comment by fxedel** [comment]
We also have such a fixture in our database:

- [JB Systems Twin Effect Laser](https://open-fixture-library.herokuapp.com/jb-systems/twin-effect-laser)

We basically use aliases for each channel to be either a working or a Nothing channel, and we added different modes for each mode that can be set in the first channel. It was quite difficult to add this fixture, but it kind of works. A solution to make this easier may be useful, but we also rate this as a low-priority thing.

**Comment by mcallegari** [comment]
Alright, took me an awful amount of time and patience, but I went through all the 9926 single capability channels and migrated them to the new format.
The space saved is significant (1.5MB on 943 definitions) and the advantages are clear.
Now the update-fixture-map.rb won't work anymore, and I know nothing about Ruby.
To be honest I'd switch to Python, which is [more popular than Ruby](https://insights.stackoverflow.com/survey/2018#technology-programming-scripting-and-markup-languages).

**Comment by janosvitok** [comment]
To be honest, I'd migrate the code to C++ (most of the code is already there) and include it in fixture editor and/or a command line tool that would be run as part of build process.

**Comment by fxedel** [comment]
Updating the [schema](https://github.com/mcallegari/qlcplus/blob/newdef/resources/schemas/fixture.xsd) should be another task in the TODO list, shouldn't it? This could also prove that your _immense_ work on updating the fixtures was right and didn't cause any errors.

As of fixture scripts: As it's important to use maintainable code, Python would probably be a good choice. (Even if I don't like it that much, but I don't know Ruby and therefore can't help updating the existing script.) I think Python is also the most commonly used scripting language in this project, isn't it?

**Comment by mcallegari** [comment]
@janosvitok  updated the definition schema. Thanks.

I've tweaked the Python script I started to generate also the Fixture map. It's now called 'fixtures-tool.py'
The idea is to have a unified tool to do everything with several command line options.
For now `--convert` and `--map` are there. `--validate` could be next.
I have also stripped the map to remove as much as possible redundant information (like the fixture file name extension). The file went from 76k to 62k.
While this doesn't make any difference on a PC, it might help less capable devices (e.g. Raspberry Pi) to start QLC+ faster. We're moving toward 1000 fixture defs, but in a long run they'll be thousands.

**Comment by mcallegari** [comment]
For who is still following this thread, I've started to add channel aliases.
As I suspected, this is a very complicated change. I already changed my mind like 3 times while writing the code. I have updated the shared document with the (hopefully) final syntax for aliases, which looks like:
```
<Capability Min="200" Max="255" Preset=”Alias”>Sound active
  <Alias Mode="10 Channel" Channel="Dimmer" With="Sound sensitivity" />
  <Alias Mode="16 Channel" Channel="Dimmer" With="Sound sensitivity" />
</Capability>
```
That you can read like this: **in** Mode="10 Channel"**, replace** Channel="Dimmer" With="Sound sensitivity" 
It's very verbose, but there are reasons behind this choice:
- aliases are stored into capabilities, but capabilities know nothing about modes and channels. So no indices can be used
- I realized that once a channel is replaced at runtime, its original name is lost, unless the whole channel name (or pointer) is stored somewhere. Therefore the "Channel" attribute is now a string and not a number, to allow "reverting" to the original channel when the DMX values exits an alias.

**Changes to the Fixture Editor**
For the above reasons, I had to move aliases at fixture level, which has knowledge of modes and all channels. To experiment, I am working on @siegmund42 Eurolite LED Bar RGB 252/10 case, which is massive in terms of alias usage.
The base mode definition looks like this:
![image](https://user-images.githubusercontent.com/2951337/39397484-2c6b8152-4b00-11e8-8fb1-862cd183517a.png)

And in the new "Alias" tab the situation is this:
![image](https://user-images.githubusercontent.com/2951337/39397487-4515ff52-4b00-11e8-8eae-0dfbfa54cb6f.png)

Obviously this is an extreme case. Normally the Alias tab will contain just 4-5 entries.
I also attach how the Eurolite bar definition looks like with all the latest changes included.
[Eurolite-LED-Bar-RGB-252_10_40__Indoor.zip](https://github.com/mcallegari/qlcplus/files/1957800/Eurolite-LED-Bar-RGB-252_10_40__Indoor.zip)

Now I "only" need to handle the channels replacement at runtime and update the UI accordingly.
God knows how many implications this carries along :cry:

**Comment by mcallegari** [comment]
There you go
![peek 2018-04-29 14-33](https://user-images.githubusercontent.com/2951337/39406657-64151bd0-4bba-11e8-8336-74596e88e838.gif)

**Comment by fxedel** [comment]
@mcallegari The preview looks awesome! :tada: 

Is it also possible to make a scene that only affects "Sound Sensitivity" and not "Program Speed", if "Sound Sensitivity" is an alias of "Program Speed"? (without setting the channel that activates this alias)

**Comment by mcallegari** [comment]
> Is it also possible to make a scene that only affects "Sound Sensitivity" and not "Program Speed", if "Sound Sensitivity" is an alias of "Program Speed"? (without setting the channel that activates this alias)

Not sure I understand the question, but if I got it right, then I think the answer is no. A Scene is just a snapshot of fixture/channel index/value triples.
In the end, aliases are just a "visual effect". The channels meaning and their usage is up to the user.

**Comment by fxedel** [comment]
Have a look at the [Chauvet DJ SlimPAR Pro QZ12](https://open-fixture-library.org/chauvet-dj/slimpar-pro-qz12) (11-channel mode):

There's a "Programs" channel that includes some Auto modes and a Sound mode and therefore switches another channel between "Program Speed" and "Sound Sensitivity".

So it makes sense to have faders for each "channel behavior", of which only one applies at a time, like so:

![bildschirmfoto von 2018-04-29 15-18-52](https://user-images.githubusercontent.com/7782229/39407083-02c000fa-4bc1-11e8-9820-abf4bdf933ce.png)
![bildschirmfoto von 2018-04-29 15-19-01](https://user-images.githubusercontent.com/7782229/39407089-0590907e-4bc1-11e8-95a3-1e7fcfea2441.png)

Having the ability to only set one channel's value (even if it may be inactive) in scenes, sequences, and so on extends this use case.

Currently, it seems like it's only possible to have a single fader for both "Program Speed" and "Sound Sensitivity", which are completely different effects (e.g. when I have loud and fast music, I want fast speed and low sensitivity). 

However, I can guess the huge effort to make changing the channel names and icons depending on an other channel's value working ;). Implementing switching channels in OFL had also been very complicated and has still some limitations.

**Comment by mcallegari** [comment]
I see the usage case, but I'd say you can resolve it by placing "Program  Speed" and "Sound Sensitivity" sliders in a multi page frame. This way both sliders control the same channel, but the value used it the one of the active frame page. Probably you can even automatically switch the frame page when clicking on VC buttons.
Makes sense ?

**Comment by fxedel** [comment]
Interesting idea! However, I wasn't able to make it working in my example show from above:

I've put each slider into a different frame page, both affect the same channel using level mode. But when I switch the page, the newly shown fader's value isn't applied – only when I move the handle a little bit with the mouse (which is not much better than having a single fader and adjusting it each time).

I thought automatic switching was easy with Loopback, but as a page is shown when the attached input channel changes, switching from "Sound" to "Auto X" didn't work: Loopback channel 1 ("Show Program Speed") and Loopback channel 2 ("Show Sound Sensitivity") have changed, and the latter takes precedence.

Considering the effort it took me to implement this in a show, I think it'll be worth saving my proposal as an idea for future additions.

**Comment by mcallegari** [comment]
@fxedel @siegmund42 @janosvitok 
I'm glad to inform you that this rework has come to an end. I've just merged the biggest QLC+ branch ever created (60+ commits) on master. And as planned, I closed #696 and #989.
Now it's "just" a matter of bugfixing the changes and improve the existing fixtures with the new features (basically 6.2 and 6.3 items)
Since this already took me an awful amount of time, I need to move on on other topics (mostly QLC+ 5) so I kindly ask you guys some help with the items above.
While migrating the existing fixtures, I've created a list of the possible definitions that could be improved (attached)
Can you please go through it and send me pull requests if you find possible improvements ?
Otherwise, adopt the method you prefer (e.g. grep) to find channels with missing information.

Thanks !

[aliases.zip](https://github.com/mcallegari/qlcplus/files/1976611/aliases.zip)

**Comment by fxedel** [comment]
@mcallegari 
Great news! Is there a rough plan when the new version will be released? We'll create a new OFL plugin for QLC+ besides the existing one soon to make the new fixture format downloadable at https://open-fixture-library.org. 

As of fixtures: I'll have a closer look at the list later. Some of them are also included in OFL, so we can replace them as soon as the new plugin is finished.

**Comment by mcallegari** [comment]
@fxedel 
the rough plan really is a QLC+ 4 release every 6 months, since it's in maintenance mode.
However, I can think of releasing a 4.12.0 one to highlight that something big has changed.

**Comment by fxedel** [comment]
@mcallegari What exactly is the difference between "PulseInSlowToFast" and "PulseOutSlowToFast"? In my understanding, pulse is a sinus or ziczac curve (of Intensity), so what's "in" and what's "out"? Maybe something like ramp up / down? If so, which would be the correct preset for a sinus pulse?

Here's a visualization of my understanding of Shutter effects:
https://jsfiddle.net/0p97phms/19/

**Comment by janosvitok** [comment]
XML schema is outdated. Fixture definitions do not validate at current git master.

**Comment by mcallegari** [comment]
> What exactly is the difference between "PulseInSlowToFast" and "PulseOutSlowToFast"? In my understanding, pulse is a sinus or ziczac curve (of Intensity), so what's "in" and what's "out"? Maybe something like ramp up / down? If so, which would be the correct preset for a sinus pulse?

@fxedel Pulse In and Pulse Out are what you call Ramp Up / Ramp Down. I tried to watch youtube videos showing this feature but I haven't found much material. And to be honest I've never seen "ramp" into definitions.

> Aliases do not validate yet. I don't like the way they are implemented. I think that a tag should contain either direct content or other tags, but not both.

@janosvitok the idea is there since February. Maybe you could come out with this statement before ?
Anyway, it's your opinion. Before chosing that way I checked, and in XML it's perfectly legit to have both values and sub tags.

**Comment by janosvitok** [comment]
> And to be honest I've never seen "ramp" into definitions, always "pulse".

* [MA2](https://help2.malighting.com/Page/grandMA2/effects_predefined/en/3.3#toc_header_anchor_2)
* [Hog4](https://www.highend.com/pub/support/controllers/documents/HTML/en/sect-table_effects.htm#figr-effect_tables)

**Comment by janosvitok** [comment]
> @janosvitok the idea is there since February. Maybe you could come out with this statement before?
I know. I didn't find the time sooner.
I've updated the PR to accept current state.

> Before chosing that way I checked, and in XML it's perfectly legit to have both values and sub tags.
The problem with that is that you can't specify text position between tags once you allow mixed content (mixed="true"). Most probably (I'm not expert on XSD) you can't write a schema that allows <A>text<B/></A> but doesn't allow <A><B/>text</A> or even <A>text1<B/>text2</A>.

**Comment by mcallegari** [comment]
So, we've got schemas for several QLC+ XML formats, but there's no documentation how to use them anywhere, nor they're used in any automation.
So, since the fixture definition validation is still missing in the current python script, I intend to implement 2 methods: one with schema validation (http://lxml.de/validation.html) and one like what I've done in PHP.
Sounds good ? Ideas ?

**Comment by fxedel** [comment]
I strongly support your proposed idea. We also have schema and logical tests in OFL and are glad everytime they find an error we wouldn't have seen. 

What about running the tests by Travis or a similar CI?

**Comment by janosvitok** [comment]
https://github.com/mcallegari/qlcplus/blob/master/resources/fixtures/check contains call to xmllint to verify the fixtures and fixturemap.xml. With little reformatting, it can be run from travis. It takes just few seconds to run. So there are two issues:
1. properly report exit codes
2. there are some failures due to missing dimensions - we need to either exclude the fixtures from testing or fill in some bogus dimensions. I guess the latter is better since the dimensions are needed for proper display.

**Comment by janosvitok** [comment]
I've implemented the above in #1065.

I propose that we split the fixtures directory by manufacturer. 1000 files in a directory is a bit much.
We may still deploy them to a single directory.

**Comment by mcallegari** [comment]
> I propose that we split the fixtures directory by manufacturer. 1000 files in a directory is a bit much.
We may still deploy them to a single directory.

I noticed that too. Mainly a GitHub limitation though...
I'd say not only to split by manufacturer (and deploy by manufacturer), but also optimize the fixture map even further, like:
```
<M n="American DJ">
  <F n="American-DJ-12P-Hex" d="12P Hex Pearl"/>
 ...
</Manufacturer>
```
We can even rename all the files since manufacturer is redundant. Like: 12P-Hex.qxf
I'll surely do the split next days. Still thinking about renaming.

**Comment by mcallegari** [comment]
Done. Fixtures are now committed and installed in folders, by manufacturer.
However, I struggled to find a way to preserve spaces on folder names (e.g. "American DJ"), but qmake doesn't help. So for now, spaces are replaced by '_' and back to spaces in the engine.
If someone is able to figure out how to properly handle spaces (in a cross platform way), please send a PR.
@janosvitok Fixture map schema needs to be updated, please

I took the chance also to improve gobos installation. The only drawback is that ANY file in folders will be installed now. But since we have control over it, only PNG files will be there.

**Comment by FloEdelmann** [comment]
Hi Massimo,

our big capability schema redesign in the Open Fixture Library is finally over, too! I was able to write a new export plugin for the new QLC+ fixture format syntax quite easily with our new syntax; see [OFL PR #547](https://github.com/OpenLightingProject/open-fixture-library/pull/547).

Please see [my comment](https://github.com/OpenLightingProject/open-fixture-library/pull/547#issuecomment-407426761) in the PR to see how the output changed from QLC+ 4 plugin to QLC+ 5 plugin for our [test fixtures](https://github.com/OpenLightingProject/open-fixture-library/blob/master/tests/test-fixtures.md).

If you have any remarks, please let us know!

**Comment by mcallegari** [comment]
@FloEdelmann well done with the export plugin ! Even though "QLC+ 5" is not exactly the correct name. New definitions can be used also on 4.12.0, when I release it. The main difference is that QLC+ 5 considers much more definition details than QLC+ 4.

So, I've dedicated the whole day to fixtures, since nobody picked up the txt file I shared almost 3 months ago. I realized some more presets were missing, so I added them.
I also added the possibility to specify the number of faces of a prism (if present, otherwise leave 0)
There is a new -optional- attribute in the `Lens` tag named `PrismFaces`.
I'm pretty confident I can simulate prisms in 3D which is super cool.

Also I moved Iris presets to the shutter category, since it seems [an iris can do also strobe/pulse effects](https://github.com/mcallegari/qlcplus/blob/master/resources/fixtures/Robe/Robe-Robin-600E-Spot.qxf#L219).

The reference definition that has pretty much everything is: [Briteq-BTX-180LS.qxf](https://github.com/mcallegari/qlcplus/blob/master/resources/fixtures/Briteq/Briteq-BTX-180LS.qxf)

In this regard, can anyone point me to some documentation or give an explanation about the differences and usages of shutter, iris, zoom and focus ? I'm interested also in the mechanical aspects of them.
For example, why should a fixture have both a shutter and an iris when the iris can do shutter effects too ?

So, except for manpower to review existing definitions capabilities and aliases, this task is pretty much over. I'll leave it open for a few more days and then close.
I will slowly review defs when I have nothing better to do.

**Comment by mcallegari** [comment]
@fxedel thanks for the info. Actually there is also the "frost" effect which is still kind of a mystery to me...
I guess it's another lens effect like a stained glass wheel or something.

Your proposal for prism faces via capability preset is good! I haven't thought about it. It might be a special case overriding the global prism faces number.
In general though, I have already understood it is almost impossible to simulate everything in 3D. Or better, very very difficult. There are too many variables on the table.
So to start I'll stick with the most common features that can be found on the market.

**Comment by janosvitok** [comment]
Frost is a (gradual) diffusion filter. It allows to change light from hard edge (spot) to soft edge (wash). Together with beam gobos (full gobo with small hole in the middle) allows to create hybrid fixtures like Robe Pointe. One fixture that is able to act as beam, spot and wash. It a spot that has frost filter to make it wash and pinhole gobo to make it a beam (although with smaller light output, since part of the light is blocked by the gobo).

(optical) Prism is a what it says: a piece of glass that splits light in several beams on its faces (think of Dark side of the moon cover). See http://www.movinghead.net/Uploads/201410/543b7337e2bf5.jpg 

The prism is either put in the light path or not, and can be rotated.

Sometimes there are more prisms (for example 3-face circular + 6-face linear), that are exclusive to each other (if they are on the same wheel/flap) or they can be selected both (if they are separate).
